url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/PricingPolicy.java

```diff
@@ -0,0 +1,457 @@
+				return this;
+			}
+			offpeakPrice = aoffpeakPrice;
+
+			for(int i = 0; i < prices.length; i++) {
+				double price = prices[i];
+				double level = alevels[i];
+				Level l = new Level(price, level);
+				levels.add(l);
+			}
+
+			for(int i = 0; i < froms.length; i++) {
+				String from = froms[i];
+				String to = tos[i];
+				Offpeak o = new Offpeak(from, to);
+				offpeaks.add(o);
+			}
+
+			return this;
+		}
+
+		/**
+		 * Builds the Pricing Policy.
+		 *
+		 * @return the pricing policy
+		 */
+		public PricingPolicy build () {
+			return new PricingPolicy(this);
+		}
+	}
+
+	/**
+	 * Instantiates a new pricing policy.
+	 *
+	 * @param builder the builder
+	 */
+	private PricingPolicy (Builder builder) {
+		type = builder.type;
+		billingCycle = builder.billingCycle;
+		fixedCharge = builder.fixedCharge;
+		contractedCapacity = builder.contractedCapacity;
+		contractedEnergy = builder.contractedEnergy;
+		energyPricing = builder.energyPricing;
+		powerPricing = builder.powerPricing;
+		fixedCost = builder.fixedCost;
+		additionalCost = builder.additionalCost;
+		offpeakPrice = builder.offpeakPrice;
+		levels = builder.levels;
+		periods = builder.periods;
+		offpeaks = builder.offpeaks;
+	}
+
+
+	/**
+	 * Gets the billing cycle duration in days.
+	 *
+	 * @return the billing cycle duration in days
+	 */
+	public int getBillingCycle() {
+		return billingCycle;
+	}
+
+	/**
+	 * Gets the fixed charge for every billing cycle.
+	 *
+	 * @return the fixed charge for every billing cycle
+	 */
+	public double getFixedCharge() {
+		return fixedCharge;
+	}
+
+	/**
+	 * Gets the type of the pricing policy.
+	 *
+	 * @return the type of the pricing policy
+	 */
+	public String getType() {
+		return type;
+	}
+
+	/**
+	 * Gets the TOU array.
+	 *
+	 * @return the TOU array
+	 */
+	public double[] getTOUArray() {
+		if(type.equalsIgnoreCase("TOUPricing")) {
+			double[] arr = new double[Constants.MIN_IN_DAY];
+			Iterator<Period> iter = periods.iterator();
+			while(iter.hasNext()) {
+				Period p = iter.next();
+				if(p.getTo().equalsIgnoreCase("00:00")) p.setTo("24:00");
+				String[] fromTokens = p.getFrom().split(":");
+				String[] toTokens = p.getTo().split(":");
+				int from = Integer.parseInt(fromTokens[0]) * 60 + Integer.parseInt(fromTokens[1]);
+				int to = Integer.parseInt(toTokens[0]) * 60 + Integer.parseInt(toTokens[1]);
+				for(int i = from; i < to; i++) {
+					arr[i] = p.getPrice();
+				}
+			}
+			return arr;
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Calculate cost.
+	 *
+	 * @param toEnergy the current energy
+	 * @param fromEnergy the previous energy
+	 * @param toEnergyOffpeak the current energy offpeak
+	 * @param fromEnergyOffpeak the previous energy offpeak
+	 * @param tick the current tick
+	 * @param maxPower the maximum power
+	 * @return the cost
+	 */
+	public double calculateCost(double toEnergy, double fromEnergy,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 63 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.